xpdan/tools.py

```diff
@@ -13,89 +13,202 @@
-        than alpha * std away from the mean are masked.
-
-        Parameters
-        ----------
-        img: 2darray
-            The  image
-        r: 2darray
-            The array which maps pixels to bins.
-            This is usually given by pyFAI.geometry.Geometry.rArray
-        alpha: float or tuple or, 1darray
-            The number of acceptable standard deviations.
-            If tuple then we use a linear distribution of alphas from alpha[0] to
-            alpha[1], if array then we use that as the distribution of alphas
-        bins: list
-            The bin edges
-        mask: 1darray
-            A starting flattened mask
-
-        Returns
-        --------
-        2darray:
-            The mask
-        """
-
-        if mask is None:
-            working_mask = np.ones(img.shape).astype(bool)
-        else:
-            working_mask = mask.copy()
-        msk_img = img[working_mask]
-        msk_r = r[working_mask]
-
-        int_r = np.digitize(r, bins[:-1], True) - 1
-        # integration
-        mean = sts.binned_statistic(msk_r, msk_img, bins=bins,
-                                    statistic='mean')[0]
-        std = sts.binned_statistic(msk_r, msk_img, bins=bins,
-                                   statistic=np.std)[0]
-        if type(alpha) is tuple:
-            alpha = np.linspace(alpha[0], alpha[1], len(std))
-        threshold = alpha * std
-        lower = mean - threshold
-        upper = mean + threshold
-
-        # single out the too low and too high pixels
-        working_mask *= img > lower[int_r]
-        working_mask *= img < upper[int_r]
-
-        return working_mask.astype(bool)
+    from xpdan.tests.utils import PDFGetterShim as PDFGetter
+from matplotlib.path import Path
+from scipy.sparse import csr_matrix
+from multiprocessing import Pool, cpu_count
+
+from skbeam.core.accumulators.binned_statistic import BinnedStatistic1D
+from skbeam.core.mask import margin, binned_outlier
+
+
+# TODO: speed this up
+def mask_ring(v_list, p_list, a_std=3.):
+    """Find outlier pixels in a single ring
+
+    Parameters
+    ----------
+    v_list: list
+        Values in ring
+    p_list: list
+        Positions in ring
+    a_std: float
+        Acceptable number of standard deviations
+
+    Returns
+    -------
+    list:
+        The positions to be masked
+    """
+    mask_list = []
+    while len(v_list) > 0:
+        norm_v_list = np.abs(v_list - np.mean(v_list)) / np.std(v_list)
+        if np.all(norm_v_list < a_std):
+            break
+        # get the index of the worst pixel
+        worst_idx = np.argmax(norm_v_list)
+        # get the position of the worst pixel
+        worst_p = p_list[worst_idx]
+        # add the worst position to the mask
+        # delete the worst position
+        v_list = np.delete(v_list, worst_idx)
+        p_list = np.delete(p_list, worst_idx)
+        mask_list.append(worst_p)
```

Comment:
instead of appending list, maybe consider creating blanck numpy array first and assigning the index